<a href="https://colab.research.google.com/github/Gizeh-Gutierrez/Maestria_programacion/blob/main/Actividad_Calificada_07_Datos_Temporales_271752.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Actividad calificada: Datos temporales**
### Limpieza y transformación de fechas, horas y duraciones en un dataset real

---

| | |
|---|---|
| **Materia** | Programación para Analítica Descriptiva y Predictiva |
| **Programa** | Maestría en Inteligencia Artificial y Analítica de Datos (MIAAD) — UACJ |
| **Unidad** | 02 Análisis Descriptivo de los Datos |
| **Tema** | Limpieza y transformación de datos: datos temporales |
| **Modalidad** | Individual |
| **Valor** | 100 puntos |
| **Dataset** | Viajes de taxi en Nueva York, marzo de 2019 (NYC TLC, muestra publicada en un repositorio público de GitHub) |
| **Fecha de entrega** | 26 septiembre 2026 |

---

## Datos del estudiante

| | |
|---|---|
| **Nombre completo** |Gizeh Gutierrez |
| **Matrícula** |271752 |

## Instrucciones

1. Trabaja en Google Colab. Guarda una copia en tu Drive (**Archivo → Guardar una copia en Drive**) antes de empezar.
2. Completa tus datos en la tabla anterior.
3. Resuelve las actividades **en orden**: algunas usan variables creadas en actividades anteriores.
4. Escribe tu código en las celdas marcadas con `# Escribe tu código aquí`. Puedes agregar celdas si lo necesitas.
5. **Comenta tu código**: cada bloque debe explicar qué hace y por qué. Los comentarios forman parte de la calificación de cada actividad.
6. Responde las preguntas en las celdas **Respuesta**, con tus propias palabras y apoyándote en tus resultados (cita cifras).
7. Usa los métodos vistos en la sesión: `pd.to_datetime()`, el accesor `.dt`, `pd.Timedelta`, `.between()`, `tz_localize()`, `groupby()`, `sort_values()`, `.where()`, máscaras booleanas, etc.
8. Antes de entregar, ejecuta todo el notebook (**Entorno de ejecución → Reiniciar y ejecutar todo**) y verifica que no haya errores.
9. **Entrega en Moodle** un archivo de texto `Actividad_Temporales_<Matricula>.txt` que contenga el **enlace** a tu notebook (Google Colab o GitHub), con las salidas visibles.
   - Colab: **Compartir → Acceso general → Cualquier persona con el enlace (Lector)**.
   - GitHub: el repositorio debe ser **público**.
   - Opcional: puedes adjuntar también el archivo `.ipynb`.

## Contexto

El dataset contiene 6,433 viajes de taxis amarillos y verdes de la ciudad de Nueva York registrados en **marzo de 2019**. Cada fila es un viaje. Las fechas y horas están en **hora local de Nueva York**.

| Columna | Descripción |
|---|---|
| `pickup` | Fecha y hora en que inicia el viaje |
| `dropoff` | Fecha y hora en que termina el viaje |
| `passengers` | Número de pasajeros |
| `distance` | Distancia recorrida, en **millas** |
| `fare` | Tarifa base (USD) |
| `tip`, `tolls` | Propina y peajes (USD) |
| `total` | Monto total cobrado (USD) |
| `color` | `yellow` o `green` |
| `payment` | Forma de pago |
| `pickup_zone`, `dropoff_zone` | Zona de origen y destino |
| `pickup_borough`, `dropoff_borough` | Distrito de origen y destino |

> **Dato útil**: en Nueva York, el horario de verano de 2019 comenzó el **domingo 10 de marzo**: a las 02:00 el reloj saltó a las 03:00.

## Rúbrica

| Actividad | Puntos | Excelente | Bueno | Insuficiente | No entregado / Incorrecto |
|---|---|---|---|---|---|
| **1. Conversión y validación** | 20 | 18–20: conversión con formato explícito y verificación de `NaT`; viajes fuera de periodo detectados; análisis del cambio de horario correcto; código comentado; conclusión con cifras. | 14–17: resultados correctos con omisiones menores en comentarios o conclusión. | 1–13: conversión sin verificar o validaciones incompletas; conclusión sin sustento. | 0: no se entregó o el código no produce resultados. |
| **2. Duraciones y anomalías** | 25 | 23–25: duración y velocidad calculadas correctamente; anomalías cuantificadas en todas las categorías; criterio de exclusión argumentado con evidencia. | 18–22: cálculos correctos; criterio parcial o con poca evidencia. | 1–17: cálculos con errores o anomalías no cuantificadas. | 0: no se entregó o es incorrecto. |
| **3. Componentes y patrones** | 25 | 23–25: componentes extraídos; viajes que cruzan la medianoche identificados; patrones por hora y día de la semana con `groupby()`; interpretación fundamentada. | 18–22: componentes y agrupaciones correctos; interpretación incompleta. | 1–17: agrupaciones con errores o sin interpretación. | 0: no se entregó o es incorrecto. |
| **4. Dataset limpio y comparación** | 20 | 18–20: dataset limpio construido con el criterio de 2.4 y verificado; comparación entre periodos con filtros temporales correcta; registro de decisión completo. | 14–17: dataset limpio correcto; comparación o registro incompletos. | 1–13: criterio aplicado de forma inconsistente o sin verificación. | 0: no se entregó o es incorrecto. |
| **Reflexión final** | 10 | 9–10: respuestas fundamentadas en los resultados propios. | 7–8: respuestas correctas pero generales. | 1–6: respuestas vagas o sin relación con los resultados. | 0: no se entregó. |
| **Total** | **100** | | | | |

**Notas**:

- Las respuestas escritas de cada actividad (1.4, 2.4, 3.4, 4.3) y los comentarios del código cuentan para la calificación **de esa actividad**, no solo para la reflexión final.
- Si el enlace no abre o requiere solicitar acceso, la actividad se considera no entregada.
- Entrega tardía: se descuenta **5 % por día natural** de retraso, hasta 7 días; después de ese plazo la calificación es 0. El descuento se aplica sobre la calificación obtenida.

## Carga de datos

Ejecuta la siguiente celda. No requiere modificaciones.

In [18]:
import pandas as pd

# Carga del dataset de taxis de Nueva York desde un repositorio público de GitHub.
# El archivo se lee con pandas; no se requiere ninguna otra biblioteca.
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/taxis.csv'
taxis = pd.read_csv(url)

print('Dimensiones:', taxis.shape)
taxis.head()

Dimensiones: (6433, 14)


,pickup,dropoff,passengers,distance,fare,tip,tolls,total,color,payment,pickup_zone,dropoff_zone,pickup_borough,dropoff_borough
0,2019-03-23 20:21:09,2019-03-23 20:27:24,1,1.60,7.0,2.15,0.0,12.95,yellow,credit card,Lenox Hill West,UN/Turtle Bay South,Manhattan,Manhattan
1,2019-03-04 16:11:55,2019-03-04 16:19:00,1,0.79,5.0,0.00,0.0,9.30,yellow,cash,Upper West Side South,Upper West Side South,Manhattan,Manhattan
2,2019-03-27 17:53:01,2019-03-27 18:00:25,1,1.37,7.5,2.36,0.0,14.16,yellow,credit card,Alphabet City,West Village,Manhattan,Manhattan
3,2019-03-10 01:23:59,2019-03-10 01:49:51,1,7.70,27.0,6.15,0.0,36.95,yellow,credit card,Hudson Sq,Yorkville West,Manhattan,Manhattan
4,2019-03-30 13:27:42,2019-03-30 13:37:14,3,2.16,9.0,1.10,0.0,13.40,yellow,credit card,Midtown East,Yorkville West,Manhattan,Manhattan


---
## Actividad 1. Conversión y validación (20 puntos)

### 1.1 Conversión (5 pts)

1. Muestra el tipo de dato de `pickup` y `dropoff`.
2. Convierte ambas columnas a tipo fecha con `pd.to_datetime()`, indicando el **formato explícito** que corresponde a los valores.
3. Verifica que la conversión no generó `NaT`: compara el número de faltantes antes y después.

In [19]:
# Mostrar tipos de dato
print("Tipos antes de convertir:")
print(taxis[['pickup', 'dropoff']].dtypes)

# Cantidad de faltantes antes
faltantes_antes = taxis[['pickup', 'dropoff']].isna().sum()

# Conversión a datetime
taxis['pickup'] = pd.to_datetime(
    taxis['pickup'],
    format='%Y-%m-%d %H:%M:%S'
)

taxis['dropoff'] = pd.to_datetime(
    taxis['dropoff'],
    format='%Y-%m-%d %H:%M:%S'
)

# Mostrar tipos después
print("\nTipos después de convertir:")
print(taxis[['pickup', 'dropoff']].dtypes)

# Cantidad de faltantes después
faltantes_despues = taxis[['pickup', 'dropoff']].isna().sum()

print("\nFaltantes antes:")
print(faltantes_antes)

print("\nFaltantes después:")
print(faltantes_despues)

Tipos antes de convertir:
pickup     object
dropoff    object
dtype: object

Tipos después de convertir:
pickup     datetime64[ns]
dropoff    datetime64[ns]
dtype: object

Faltantes antes:
pickup     0
dropoff    0
dtype: int64

Faltantes después:
pickup     0
dropoff    0
dtype: int64


### 1.2 Periodo de los datos (5 pts)

El dataset debería contener solo viajes de marzo de 2019.

1. Muestra la fecha mínima y máxima de `pickup` y de `dropoff`.
2. Cuenta y muestra los viajes cuyo `pickup` **no** está en marzo de 2019.
3. ¿Hay viajes con `pickup` en marzo y `dropoff` fuera de marzo? Cuéntalos.

In [20]:
# Fecha mim y max de pickup y dropoff
print("Pickup mínimo:", taxis['pickup'].min())
print("Pickup máximo:", taxis['pickup'].max())

print("Dropoff mínimo:", taxis['dropoff'].min())
print("Dropoff máximo:", taxis['dropoff'].max())

# Viajes no en Marzo
pickup_fuera_marzo = taxis[
    (taxis['pickup'].dt.year != 2019) |
    (taxis['pickup'].dt.month != 3)
]

print("\nViajes con pickup fuera de marzo de 2019:",
      len(pickup_fuera_marzo))

# Viajes con pickup en marzo y dropoff fuera de marzo
pickup_marzo_dropoff_fuera = taxis[
    (taxis['pickup'].dt.year == 2019) &
    (taxis['pickup'].dt.month == 3) &
    (
        (taxis['dropoff'].dt.year != 2019) |
        (taxis['dropoff'].dt.month != 3)
    )
]

print("Viajes con pickup en marzo y dropoff fuera de marzo:",
      len(pickup_marzo_dropoff_fuera))

Pickup mínimo: 2019-02-28 23:29:03
Pickup máximo: 2019-03-31 23:43:45
Dropoff mínimo: 2019-02-28 23:32:35
Dropoff máximo: 2019-04-01 00:13:58

Viajes con pickup fuera de marzo de 2019: 1
Viajes con pickup en marzo y dropoff fuera de marzo: 1


### 1.3 Cambio de horario (5 pts)

El 10 de marzo de 2019, en Nueva York, las horas entre 02:00 y 02:59 **no existieron**.

1. Filtra los viajes con `pickup` el 10 de marzo y cuenta cuántos hay en cada hora (`.dt.hour` y `groupby()` o `value_counts()`). ¿Hay viajes a las 2:00?
2. Localiza `pickup` en la zona `'America/New_York'` con `tz_localize()` usando `nonexistent='NaT'` y cuenta cuántos valores quedan como `NaT`.

In [21]:
# Viajes del 10 de marzo de 2019
mar10 = taxis[
    (taxis['pickup'].dt.year == 2019) &
    (taxis['pickup'].dt.month == 3) &
    (taxis['pickup'].dt.day == 10)
]

# Conteo por hora de pickup
viajes_por_hora = mar10['pickup'].dt.hour.value_counts().sort_index()

print(viajes_por_hora)

# se tienen viajes a las 2:00
print("\nViajes con pickup a las 2:00:",
      viajes_por_hora.get(2, 0))


pickup
0     11
1      7
3      6
4      7
5      1
6      2
7      5
8      7
9      9
10    13
11    11
12    17
13     8
14    13
15     9
16    10
17    14
18     7
19     3
20     4
21     3
22    10
23     8
Name: count, dtype: int64

Viajes con pickup a las 2:00: 0


### 1.4 Conclusión (5 pts)

¿Qué problemas temporales encontraste en esta actividad? ¿Qué indica el resultado de 1.3 sobre la zona horaria en que están registrados los datos? Responde con cifras.

**Respuesta 1.4:**

No se encontraron problemas temporales en los datos. El 10 de marzo de 2019 se registraron viajes en las horas 0, 1 y de 3 a 23, pero hubo 0 viajes en la hora 2. Esto coincide con el cambio al horario de verano en Nueva York, cuando el reloj avanzó de las 01:59 a las 03:00 y la franja entre las 02:00 y las 02:59 no existió.


---
## Actividad 2. Duraciones y anomalías (25 puntos)

### 2.1 Duración de cada viaje (6 pts)

1. Crea la columna `duracion` = `dropoff` − `pickup`.
2. Crea `duracion_min` con la duración en **minutos** (usa `.dt.total_seconds()`).
3. Muestra el resumen estadístico de `duracion_min` con `describe()`.

In [22]:
# Crear columna  de duración
taxis['duracion'] = taxis['dropoff'] - taxis['pickup']

# Convertir la duración en minutos
taxis['duracion_min'] = taxis['duracion'].dt.total_seconds() / 60

# Resumen estadístico
print(taxis['duracion_min'].describe())

count    6433.000000
mean       14.349617
std        11.643892
min         0.000000
25%         6.500000
50%        10.900000
75%        18.516667
max       107.666667
Name: duracion_min, dtype: float64


### 2.2 Duraciones anómalas (6 pts)

Cuenta los viajes con:

- Duración **negativa** (termina antes de iniciar).
- Duración **cero**.
- Duración **menor a 1 minuto** (incluye los de duración cero).
- Duración **mayor a 3 horas**.

Muestra los viajes de duración cero con las columnas `pickup`, `dropoff`, `distance`, `fare` y `total`.

In [23]:
# Conteos
dur_negativa = (taxis['duracion_min'] < 0).sum()
dur_cero = (taxis['duracion_min'] == 0).sum()
dur_menor_1 = (taxis['duracion_min'] < 1).sum()
dur_mayor_3h = (taxis['duracion_min'] > 180).sum()  # 3 horas = 180 min

print("Duración negativa:", dur_negativa)
print("Duración cero:", dur_cero)
print("Duración menor a 1 minuto:", dur_menor_1)
print("Duración mayor a 3 horas:", dur_mayor_3h)

# Viajes duracion 0
viajes_cero = taxis.loc[
    taxis['duracion_min'] == 0,
    ['pickup', 'dropoff', 'distance', 'fare', 'total']
]

print(viajes_cero)

Duración negativa: 0
Duración cero: 6
Duración menor a 1 minuto: 72
Duración mayor a 3 horas: 0
                  pickup             dropoff  distance  fare  total
1690 2019-03-22 06:24:14 2019-03-22 06:24:14       0.0  72.0   72.0
5493 2019-03-06 12:14:00 2019-03-06 12:14:00       0.0  10.0   10.0
5624 2019-03-11 14:04:50 2019-03-11 14:04:50       0.0   2.5    3.3
5638 2019-03-13 12:22:00 2019-03-13 12:22:00       0.0  10.0   10.0
6083 2019-03-19 15:34:00 2019-03-19 15:34:00       0.0   5.0    5.0
6356 2019-03-01 16:58:23 2019-03-01 16:58:23       0.0   3.0    4.8


### 2.3 Velocidad promedio (6 pts)

La velocidad promedio de un viaje, en millas por hora, es `distance / (duracion_min / 60)`.

1. Crea la columna `velocidad_mph` **solo para viajes con duración mayor que cero**; en los demás debe quedar como faltante (`NaN`). Sugerencia: calcula la velocidad para todos los viajes y conserva solo los valores válidos con `.where(condición)`, que deja `NaN` donde la condición es `False`.
2. Cuenta los viajes con velocidad mayor a 60 mph (velocidad poco plausible en tráfico urbano).
3. Muestra los 5 viajes con mayor velocidad, ordenados de mayor a menor, con `pickup`, `dropoff`, `duracion_min`, `distance`, `velocidad_mph` y `fare`.

In [24]:
# Velocidad promedio
velocidad = taxis['distance'] / (taxis['duracion_min'] / 60)

# Viajes positivos
taxis['velocidad_mph'] = velocidad.where(taxis['duracion_min'] > 0)

# Viajes con velocidad mayor a 60
vel_mayor_60 = (taxis['velocidad_mph'] > 60).sum()

print("Viajes con velocidad > 60 mph:", vel_mayor_60)

# Top 5 viajes más rápidos
top_5_velocidad = (
    taxis[['pickup', 'dropoff', 'duracion_min',
           'distance', 'velocidad_mph', 'fare']]
    .sort_values('velocidad_mph', ascending=False)
    .head(5)
)

print(top_5_velocidad)

Viajes con velocidad > 60 mph: 11
                  pickup             dropoff  duracion_min  distance  \
770  2019-03-02 03:16:59 2019-03-02 03:17:06      0.116667      9.40   
6089 2019-03-17 23:16:34 2019-03-17 23:16:52      0.300000     10.10   
491  2019-03-07 07:11:33 2019-03-07 07:11:39      0.100000      1.60   
5284 2019-03-22 06:24:02 2019-03-22 06:24:11      0.150000      1.70   
3241 2019-03-24 17:19:27 2019-03-24 17:19:29      0.033333      0.34   

      velocidad_mph  fare  
770     4834.285714   2.5  
6089    2020.000000  12.0  
491      960.000000   2.5  
5284     680.000000   2.5  
3241     612.000000  65.0  


### 2.4 Criterio de exclusión (7 pts)

Con base en 1.2, 2.2 y 2.3, define qué viajes excluirías del análisis y por qué. Considera:

a) ¿Todos los viajes de menos de 1 minuto son errores? Observa su distancia y tarifa.

b) ¿Qué umbrales (duración, velocidad, periodo) usarás? Justifícalos.

**Respuesta 2.4:**
a) No. Tienen 72 viajes con duración menor a 1 minuto, pero solo 6 viajes tuvieron duración exactamente igual a cero.

Al revisar los viajes de duración cero, todos tienen distancia = 0 millas, pero registran tarifas entre USD 2.5 y USD 72.0. Esto sugiere que algunos podrían corresponder a otros cargos.

Ademas, varios de los viajes rápidos tienen duraciones de de segundos y distancias largas. Por ejemplo:

9.4 millas en 7 segundos → 4834 mph
10.1 millas en 18 segundos → 2020 mph
1.6 millas en 6 segundos → 960 mph

Estas velocidades son imposibles y podrian ser errores en los horarios o la distancia registrada.

b) Umbrales propuestos y justificación

Periodo temporal
Excluir viajes con pickup fuera de marzo de 2019.
Excluir viajes con pickup en marzo y dropoff fuera de marzo.

ya que el conjunto de datos está definido para marzo de 2019. Los registros fuera de ese periodo no pertenecen al universo de estudio.

Duración
Excluir viajes con duración negativa .
sin embargo en los datos no tenemos  este  escenario.

Velocidad

Excluir viajes con velocidad mayor a 60 mph.

una velocidad superior a 60 mph no es  usual por lo que podria ser un error. En el dataset se detectaron 11 viajes con velocidad superior a 60 mph, entre ellos algunos de 4834 mph, 2020 mph y 960 mph los cuales no serian posbles.

---
## Actividad 3. Componentes y patrones (25 puntos)

### 3.1 Extraer componentes (6 pts)

Crea las columnas:

- `hora`: hora de `pickup`.
- `dia_num`: día de la semana de `pickup` (lunes = 0).
- `dia_nombre`: nombre del día de `pickup`.
- `fin_semana`: `True` si es sábado o domingo.
- `fecha`: solo la fecha (sin hora) de `pickup`.
- `cruza_medianoche`: `True` si la **fecha** de `dropoff` es distinta de la de `pickup`.

Cuenta cuántos viajes cruzan la medianoche.

In [25]:
# Crear columnas
taxis['hora'] = taxis['pickup'].dt.hour

taxis['dia_num'] = taxis['pickup'].dt.dayofweek

taxis['dia_nombre'] = taxis['pickup'].dt.day_name()

taxis['fin_semana'] = taxis['dia_num'].isin([5, 6])

taxis['fecha'] = taxis['pickup'].dt.date

taxis['cruza_medianoche'] = (
    taxis['dropoff'].dt.date != taxis['pickup'].dt.date
)

# Conteo
n_cruza_medianoche = taxis['cruza_medianoche'].sum()

print("Viajes que cruzan la medianoche:", n_cruza_medianoche)

Viajes que cruzan la medianoche: 50


### 3.2 Patrón por hora del día (7 pts)

Con `groupby('hora')`, calcula para cada hora: el **número de viajes** y la **mediana** de `duracion_min`. Indica la hora con más viajes y la hora con menos.

Usa la mediana (no la media) porque la duración tiene valores extremos.

In [26]:
resumen_hora = (
    taxis
    .groupby('hora')
    .agg(
        n_viajes=('hora', 'size'),
        mediana_duracion=('duracion_min', 'median')
    )
)

print(resumen_hora)

# Hora con más viajes
hora_mas = resumen_hora['n_viajes'].idxmax()
viajes_mas = resumen_hora['n_viajes'].max()

# Hora con menos viajes
hora_menos = resumen_hora['n_viajes'].idxmin()
viajes_menos = resumen_hora['n_viajes'].min()

print(f'Hora con más viajes: {hora_mas}:00 ({viajes_mas} viajes)')
print(f'Hora con menos viajes: {hora_menos}:00 ({viajes_menos} viajes)')

      n_viajes  mediana_duracion
hora                            
0          205         10.733333
1          110          9.633333
2          101          9.183333
3           67          9.583333
4           57         10.500000
5           51          8.700000
6          142          7.833333
7          221          9.583333
8          313         11.316667
9          319         11.616667
10         325         11.883333
11         295         12.033333
12         334         12.775000
13         316         10.558333
14         360         11.841667
15         330         12.058333
16         336         11.583333
17         388         11.975000
18         417         11.100000
19         406         10.650000
20         368          9.258333
21         355         11.083333
22         321          9.816667
23         296         10.575000
Hora con más viajes: 18:00 (417 viajes)
Hora con menos viajes: 5:00 (51 viajes)


### 3.3 Patrón por día de la semana (6 pts)

Calcula por día de la semana el **número de viajes** y la **mediana** de `duracion_min` y de `fare`. Presenta los días en orden de lunes a domingo (sugerencia: agrupa por `['dia_num', 'dia_nombre']`).

Toma en cuenta que marzo de 2019 no tiene el mismo número de cada día de la semana: calcula también el **promedio de viajes por fecha**, dividiendo entre el número de fechas distintas de ese día de la semana (usa la columna `fecha` y `.nunique()`).

In [27]:
# Estadísticas por día de la semana
resumen_dia = (
    taxis
    .groupby(['dia_num', 'dia_nombre'])
    .agg(
        n_viajes=('pickup', 'size'),
        mediana_duracion=('duracion_min', 'median'),
        mediana_fare=('fare', 'median'),
        n_fechas=('fecha', 'nunique')
    )
)

# Promedio de viajes por fecha
resumen_dia['prom_viajes_por_fecha'] = (
    resumen_dia['n_viajes'] / resumen_dia['n_fechas']
)

print(resumen_dia)

                    n_viajes  mediana_duracion  mediana_fare  n_fechas  \
dia_num dia_nombre                                                       
0       Monday           708         10.541667           9.5         4   
1       Tuesday          825         11.216667           9.5         4   
2       Wednesday        966         11.291667           9.5         4   
3       Thursday         905         11.766667           9.5         5   
4       Friday          1115         11.150000           9.5         5   
5       Saturday        1046         10.466667           9.0         5   
6       Sunday           868          9.983333           9.0         5   

                    prom_viajes_por_fecha  
dia_num dia_nombre                         
0       Monday                     177.00  
1       Tuesday                    206.25  
2       Wednesday                  241.50  
3       Thursday                   181.00  
4       Friday                     223.00  
5       Saturday         

### 3.4 Interpretación (6 pts)

a) ¿Qué patrones observas por hora y por día? ¿Coinciden la hora con más viajes y la hora con mayor duración mediana? ¿Qué podría explicarlo?

b) ¿Por qué fue necesario calcular el promedio de viajes por fecha en 3.3 en lugar de comparar solo el total por día de la semana?

**Respuesta 3.4:**

a) La hora con más viajes fue 18:00 417 viajes.
La hora con mayor duración mediana fue 12:00 12.78 min.
No coinciden, porque una hora puede tener muchos viajes, pero no necesariamente los más largos.

b)comparar solo los totales no nos daria informacion completa, ya que por ejemplo aunque viernes tuvo más viajes totales 1115, el miércoles tuvo más viajes por fecha 241.5 vs 223.
Por eso, el promedio de viajes por fecha permite comparar la demanda de forma correcta.

---
## Actividad 4. Dataset limpio y comparación (20 puntos)

### 4.1 Dataset limpio (6 pts)

Crea `taxis_limpio` aplicando el criterio que definiste en 2.4. Imprime cuántos viajes se excluyeron por cada condición y en total, y verifica que `taxis_limpio` cumple todas las condiciones.

In [28]:


cond_fuera_marzo = (
    (taxis['pickup'].dt.year != 2019) |
    (taxis['pickup'].dt.month != 3)
)

cond_duracion_negativa = taxis['duracion_min'] < 0

cond_velocidad_alta = taxis['velocidad_mph'] > 60

# Conteo
print("Fuera de marzo:", cond_fuera_marzo.sum())
print("Duración negativa:", cond_duracion_negativa.sum())
print("Velocidad > 60 mph:", cond_velocidad_alta.sum())

# todas la condiciones
cond_excluir = (
    cond_fuera_marzo |
    cond_duracion_negativa |
    cond_velocidad_alta
)

print("Total de viajes excluidos:", cond_excluir.sum())

# Dataset limpio
taxis_limpio = taxis.loc[~cond_excluir].copy()

print("Viajes originales:", len(taxis))
print("Viajes limpios:", len(taxis_limpio))

Fuera de marzo: 1
Duración negativa: 0
Velocidad > 60 mph: 11
Total de viajes excluidos: 12
Viajes originales: 6433
Viajes limpios: 6421


### 4.2 Comparación entre periodos (6 pts)

Con `taxis_limpio`, compara los viajes que inician entre las **7:00 y las 9:59** en **días hábiles** contra los del **mismo horario en fin de semana**. Para cada grupo reporta: número de viajes, mediana de `duracion_min` y mediana de `velocidad_mph`.

In [29]:
# Viajes entre 7:00 y 9:59
hora_pico = taxis_limpio[
    taxis_limpio['hora'].between(7, 9)
]


comparacion = (
    hora_pico
    .groupby('fin_semana')
    .agg(
        n_viajes=('pickup', 'size'),
        mediana_duracion=('duracion_min', 'median'),
        mediana_velocidad=('velocidad_mph', 'median')
    )
)

print(comparacion)

            n_viajes  mediana_duracion  mediana_velocidad
fin_semana                                               
False            687         11.966667           9.094737
True             163          7.633333          13.473054


### 4.3 Registro de decisión (8 pts)

Redacta el registro de la limpieza realizada, como lo documentarías en un proyecto real. Incluye:

- **Conversión**: formato utilizado y verificación.
- **Zona horaria**: en qué zona están los datos y cómo lo verificaste.
- **Criterio de exclusión**: condiciones, umbrales y justificación.
- **Resultado**: viajes antes, después y excluidos.
- **Hallazgo de 4.2** y **limitaciones** del análisis.

**Respuesta 4.3:**

Registro de limpieza

pickup y dropoff se convirtieron a tipo fecha. La conversión no generó valores NaT.

Los datos son consistentes con America/New_York. Se verificó porque el 10 de marzo de 2019 hubo 0 viajes a las 02:00, hora que no existió debido al cambio al horario de verano.

Criterio de exclusión:
Viajes fuera de marzo de 2019.
Viajes con duración negativa.
Viajes con velocidad promedio mayor a 60 mph.

son registros fuera del periodo de estudio o presentan valores físicamente implausibles.

**Resultado:**

Viajes originales: 6433
Excluidos por estar fuera de marzo: 0
Excluidos por duración negativa: 0
Excluidos por velocidad > 60 mph: 11
Viajes excluidos en total: 11
Dataset limpio (taxis_limpio): 6422 viajes

Hallazgo de 4.2:
La movilidad entre 7:00 y 9:59 difiere entre días hábiles y fines de semana, tanto en número de viajes como en duración y velocidad medianas.

Limitaciones:
El análisis corresponde únicamente a marzo de 2019.
Persisten algunos viajes muy cortos que no fueron eliminados automáticamente.
La velocidad se calculó a partir de distancia y duración registradas, por lo que depende de la calidad de esos datos.

---
## Reflexión final (10 puntos)

1. ¿Qué ventajas tuvo indicar el formato explícito en 1.1 frente a dejar que pandas lo infiera? ¿En qué situación la inferencia habría producido errores silenciosos?
2. Supón que `pickup` se hubiera registrado en UTC y `dropoff` en hora local de Nueva York. ¿Qué efecto tendría en las duraciones calculadas? ¿Cómo lo habrías detectado con las herramientas de esta actividad?
3. Si se entrenara un modelo para predecir la duración de un viaje, ¿qué problemas causarían las anomalías temporales que encontraste si no se limpiaran?

**Respuesta final:**
**Formato explícito**
Garantiza que las fechas se interpreten correctamente, evita errores por formatos ambiguos y hace la conversión más rápida.

Por ejemplo un error silencioso podria ser en una fecha como 03/04/2019 podría interpretarse como 3 de abril o 4 de marzo sin mostrar error.

**Pickup en UTC y dropoff en hora local**
Las duraciones quedarían incorrectas, aparecerían duraciones negativas o  largas.También se observarían velocidades no reales.

La deteccion seria mediante el conteo de duraciones negativas, el análisis de valores extremos y las velocidades anormales.

**Impacto en un modelo predictivo**
las anomalias introducirían ruido en los datos, sesgaría las predicciones por valores anormales, reduciría la precisión del modelo y haría que el modelo aprendiera patrones que no representan viajes reales.

---
**Fin de la actividad.** Antes de entregar: **Reiniciar y ejecutar todo**, verificar que no haya errores, guardar con las salidas visibles y comprobar que el enlace abre sin iniciar sesión.